feat(BA-3209): Add `artifact_storages` common table

jopemachine opened the pull request:
resolves #7040 (BA-3209)
<!-- replace NNN, MMM with the GitHub issue number and the corresponding Jira issue number. -->

<!--
Please precisely, concisely, and concretely describe what this PR changes, the rationale behind codes,
and how it affects the users and other developers.
-->

**Checklist:** (if applicable)

- [x] Mention to the original issue

<!-- readthedocs-preview sorna start -->
----
📚 Documentation preview 📚: https://sorna--7057.org.readthedocs.build/en/7057/

<!-- readthedocs-preview sorna end -->

<!-- readthedocs-preview sorna-ko start -->
----
📚 Documentation preview 📚: https://sorna-ko--7057.org.readthedocs.build/ko/7057/

<!-- readthedocs-preview sorna-ko end -->

seedspirit commented on src/ai/backend/manager/models/vfs_storage/row.py:
```diff
@@ -43,23 +49,27 @@ class VFSStorageRow(Base):
         primaryjoin=_get_vfs_storage_association_artifact_join_cond,
         overlaps="association_artifacts_storages_rows,object_storage_row",
     )
+    meta = relationship(
```
(Not necessarily required) How about changing the name to `descriptor` or `identity` instead of `meta`? It seems there's potential for confusion with SQLAlchemy's `MetaData`.

jopemachine commented:
I will request PR review after release 26.2.0

jopemachine committed:
feat: Add `artifact_storages` table

- Add artifact_storages table and migration script
- Move name column to artifact_storages
- Add proper selectinload in db_source
- Add ID type and update version
- Append admin prefix
- Fix rename types and error domain
- Add news fragment

jopemachine commented on src/ai/backend/common/types.py:
```diff
@@ -306,6 +306,10 @@ def check_typed_tuple(value: tuple[Any, ...], types: tuple[type, ...]) -> tuple[
+# ID of the `artifact_storages` common table (storage metadata).
+ArtifactStorageId = NewType("ArtifactStorageId", UUID)
+# ID of the concrete storage table (`object_storages`, `vfs_storages`, etc.).
+ConcreteStorageId = NewType("ConcreteStorageId", UUID)
```
I wiil rename it `ConcreteArtifactStorageId`

jopemachine committed:
fix: Reflect feedback

jopemachine commented on src/ai/backend/manager/models/artifact_storages.py:
```diff
@@ -0,0 +1,84 @@
+
+    def to_dataclass(self) -> ArtifactStorageData:
+        from ai.backend.common.types import ArtifactStorageId, ConcreteStorageId
+        from ai.backend.manager.data.artifact_storages.types import ArtifactStorageData
```
I will move those types to common, then we don't need them there

jopemachine committed:
fix: alembic conflcit

fregataa reviewed (commented):
I think that using SQLAlchemy Table inheritance would be great here.

fregataa commented on src/ai/backend/manager/repositories/vfs_storage/db_source/db_source.py:
```diff
@@ -28,42 +33,87 @@ async def get_by_name(self, storage_name: str) -> VFSStorageData:
+            # Set the storage_id on the meta creator spec and create ArtifactStorageRow
+            meta_spec = meta_creator.spec
+            if not isinstance(meta_spec, ArtifactStorageCreatorSpec):
+                raise InternalServerError("meta_creator.spec must be ArtifactStorageCreatorSpec")
```
Let's avoid the need of such check

jopemachine committed:
fix: Apply SQLAlchemy Joined Table Inheritance to artifact_storages (experimetal)

jopemachine committed:
chore: update api schema dump

Co-authored-by: octodog <mu001@lablup.com>

jopemachine committed:
fix: Migration script

jopemachine committed:
wip

jopemachine committed:
wip

jopemachine committed:
test: Fix Broken tests

fregataa reviewed (commented):
Are there storage insert/update unit tests? It would be good to have them.
Plus, how do you think of [Single Table Inheritance](https://docs.sqlalchemy.org/en/21/orm/inheritance.html#single-table-inheritance)? Update and insert execution got complicated because of joined table inheritance, so I guess single table one makes it easy

jopemachine commented:
> Are there storage insert/update unit tests? It would be good to have them. Plus, how do you think of [Single Table Inheritance](https://docs.sqlalchemy.org/en/21/orm/inheritance.html#single-table-inheritance)? Update and insert execution got complicated because of joined table inheritance, so I guess single table one makes it easy

I believe STI would not good for extensibility because it will introduce many nullable columns.

jopemachine committed:
test: Add

jopemachine commented:
Since this PR has gone stale, I'll close it. This issue should be addressed when we redesign the storage proxy after the agent refactoring is complete.
